src/oidc_child/oidc_child.c

```diff
@@ -0,0 +1,571 @@
+                           && (opts->issuer_url != NULL
+                                || opts->issuer != NULL))
+        || (opts->issuer_url == NULL && opts->issuer == NULL
+                && ((opts->device_auth_endpoint != NULL
+                        && opts->token_endpoint == NULL)
+                   || (opts->device_auth_endpoint == NULL
+                        && opts->token_endpoint != NULL)))
+        || (opts->issuer_url == NULL && opts->issuer == NULL
+                && (opts->device_auth_endpoint == NULL
+                        || opts->token_endpoint == NULL))) {
+        fprintf(stderr, "\n--issuer-url or --issuer or --device-auth-endpoint "
+                        "together with --token-endpoint are mutually exclusive "
+                        "but one variant must be given.\n\n");
+        goto done;
+    }
+
+    if (opts->client_id == NULL) {
+        fprintf(stderr, "\n--client-id must be given.\n\n");
+        goto done;
+    }
+
+    poptFreeContext(pc);
+    print_usage = false;
+
+    debug_prg_name = talloc_asprintf(NULL, "oidc_child[%d]", getpid());
+    if (debug_prg_name == NULL) {
+        ERROR("talloc_asprintf failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    opts->opt_logger = opt_logger;
+
+    if (debug_fd != -1) {
+        opts->opt_logger = sss_logger_str[FILES_LOGGER];
+        ret = set_debug_file_from_fd(debug_fd);
+        if (ret != EOK) {
+            opts->opt_logger = sss_logger_str[STDERR_LOGGER];
+            ERROR("set_debug_file_from_fd failed.\n");
+            return ret;
+        }
+    }
+
+    ret = EOK;
+
+done:
+    if (print_usage) {
+        poptPrintUsage(pc, stderr, 0);
+        poptFreeContext(pc);
+    }
+
+    return ret;
+}
+
+void trace_device_code(struct devicecode_ctx *dc_ctx, bool get_device_code)
+{
+    if (!DEBUG_IS_SET(SSSDBG_TRACE_ALL)) {
+        return;
+    }
+
+    if (get_device_code) {
+        DEBUG(SSSDBG_TRACE_ALL, "user_code: [%s].\n", dc_ctx->user_code);
+        DEBUG(SSSDBG_TRACE_ALL, "verification_uri: [%s].\n",
+                                dc_ctx->verification_uri);
+        DEBUG(SSSDBG_TRACE_ALL, "verification_uri_complete: [%s].\n",
+                                dc_ctx->verification_uri_complete == NULL ? "-"
+                                           : dc_ctx->verification_uri_complete);
+        DEBUG(SSSDBG_TRACE_ALL, "message: [%s].\n", dc_ctx->message);
+    }
+    DEBUG(SSSDBG_TRACE_ALL, "device_code: [%s].\n", dc_ctx->device_code);
+    DEBUG(SSSDBG_TRACE_ALL, "expires_in: [%d].\n", dc_ctx->expires_in);
+    DEBUG(SSSDBG_TRACE_ALL, "interval: [%d].\n", dc_ctx->interval);
+}
+
+void trace_tokens(struct devicecode_ctx *dc_ctx)
+{
+    if (!DEBUG_IS_SET(SSSDBG_TRACE_ALL)) {
+        return;
+    }
+
+    if (dc_ctx->access_token_payload != NULL) {
+        DEBUG(SSSDBG_TRACE_ALL, "access_token payload: [%s].\n",
+              json_dumps(dc_ctx->access_token_payload, 0));
```

Comment:
This (and similar calls) leak memory allocated by `json_dumps()`

Reply:
fixed